### PR TITLE
Adds aggregateWithWildcards (and sum / multiply / avg)

### PR DESCRIPTION
**Pull request by robert-milan**
This PR adds aggregateWithWildcards and sumSeriesWithWildcards, multiplySeriesWithWildcards, and averageSeriesWithWildcards

I have tested it locally against sumSeries, averageSeries, and multiplySeries and it appears to be working properly.

### CODE REVIEW

**Comment by Dieterbe** [comment]
in docker-dev, issuing this query `averageSeriesWithWildcards(perSecond(metrictank.stats.docker-env.default.api.request.*.status.*.counter32), 6)`

on graphite, I get these 2 series:
```
perSecond(metrictank.stats.docker-env.default.api.request.status.200.counter32)
perSecond(metrictank.stats.docker-env.default.api.request.status.404.counter32)
```

on MT, I get:
```
averageSeries(metrictank.stats.docker-env.default.api.request.status.200.counter32)
averageSeries(perSecond(metrictank.stats.docker-env.default.api.request.*.status.*.counter32))
```

File: expr/parse.go [hunk 1]
```diff
@@ -476,3 +476,38 @@ func aggKey(serie models.Series, nodes []expr) string {
...
+
+// filterNodesByPositions removes the node at the given position(s)
+func filterNodesByPositions(serie models.Series, nodes []expr) (string, error) {
+	metric := extractMetric(serie.Target)
+	if len(metric) == 0 {
+		return "", fmt.Errorf("invalid metric name")
+	}
+	// Trim off tags (if they are there) and split on '.'
+	parts := strings.Split(strings.SplitN(metric, ";", 2)[0], ".")
+	name := parts[:0]
```
Comment (Dieterbe) [inline expr/parse.go:488]: Why do you have all this? Graphite doesn't do this.
Graphite does only `key = '.'.join(map(lambda x: x[1], filter(lambda i: i[0] not in positions, enumerate(series.name.split('.')))))`
We need to be compatible with graphite.

File: expr/parse.go [hunk 2]
```diff
@@ -476,3 +476,38 @@ func aggKey(serie models.Series, nodes []expr) string {
...
+		// if we have already reached the end of the nodes to filter, just add all remaining parts to name
+		if n >= len(nodes) {
+			name = append(name, p)
+			continue
+		}
+
+		// we don't perform a check of ettype here. This should be done before calling this function
+		idx := int(nodes[n].int)
+		if idx >= len(parts) || idx < 0 {
+			return "", fmt.Errorf("invalid node: %d", idx)
```
Comment (Dieterbe) [inline expr/parse.go:501]: graphite doesn't do this. it allows you to specify positions that don't apply.

File: expr/parse.go [hunk 3]
```diff
@@ -476,3 +476,38 @@ func aggKey(serie models.Series, nodes []expr) string {
...
+
+	n := 0
+	for i, p := range parts {
+		// if we have already reached the end of the nodes to filter, just add all remaining parts to name
+		if n >= len(nodes) {
+			name = append(name, p)
+			continue
+		}
+
+		// we don't perform a check of ettype here. This should be done before calling this function
```
Comment (Dieterbe) [inline expr/parse.go:498]: assumptions like this are better written in the function documentation such that it's more obvious to the caller

File: expr/parse.go [hunk 4]
```diff
@@ -476,3 +476,38 @@ func aggKey(serie models.Series, nodes []expr) string {
...
+		idx := int(nodes[n].int)
+		if idx >= len(parts) || idx < 0 {
+			return "", fmt.Errorf("invalid node: %d", idx)
+		}
+
+		// if our idx matches the current position, incremenmt and continue
```
Comment (Dieterbe) [inline expr/parse.go:504]: positions may not be specified in order. we have to support that just like graphite does

Comment (robert-milan) [reply expr/parse.go:504]: We can just sort them before doing any of this

File: expr/func_aggregatewithwildcards_test.go [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+}
+
+func TestAggregateWithWildcardsIdentity(t *testing.T) {
+	testAggregateWithWildcards(
+		"identity",
+		"average",
+		[]models.Series{
+			getSeriesNamed("single", a),
+		},
+		getSeriesNamed("averageSeries(single)", a),
```
Comment (Dieterbe) [inline expr/func_aggregatewithwildcards_test.go:24]: graphite doesn't add the wrapping "averageSeries(...)" when using aggregateWithWildcards, so I think neither should we

**Comment by Dieterbe** [comment]
@robert-milan I added some commits. now it seems to work like graphite in all my tests.
I think it's good now. what do you think?
I have used this dashboard for testing: https://gist.github.com/Dieterbe/f7f9375aafa2173923c5250224f73e7e

**Comment by shanson7** [comment]
FWIW I used https://github.com/shanson7/mtcmptest to compare MT and graphite responses

**Comment by robert-milan** [comment]
Thanks Dieter, looks good.

**Pull request merged**